### PR TITLE
[SDK-3735] Create tree-shakable `Crypto` module

### LINKED ISSUE #1396: Create tree-shakable `Crypto` module
- depends on #1374 
- create new `Crypto` class which, if provided to `BaseClient` constructor, allows encrypted channels to be used

### PR DESCRIPTION
**Pull request by lawrence-forooghian**
**Note: This PR is based on top of #1417; please review that one first.**

This encapsulates the crypto functionality in a tree-shakable module. See commit messages for more details.

```
BaseRest: 87.17 KiB
BaseRest + Rest: 87.17 KiB
BaseRest + Crypto: 90.70 KiB
BaseRest + generateRandomKey: 90.73 KiB
BaseRest + getDefaultCryptoParams: 90.73 KiB
BaseRest + decodeMessage: 87.53 KiB
BaseRest + decodeEncryptedMessage: 91.05 KiB
BaseRest + decodeMessages: 87.64 KiB
BaseRest + decodeEncryptedMessages: 91.16 KiB
BaseRealtime: 144.00 KiB
BaseRealtime + Rest: 152.65 KiB
BaseRealtime + Crypto: 147.53 KiB
BaseRealtime + generateRandomKey: 147.56 KiB
BaseRealtime + getDefaultCryptoParams: 147.56 KiB
BaseRealtime + decodeMessage: 144.36 KiB
BaseRealtime + decodeEncryptedMessage: 147.88 KiB
BaseRealtime + decodeMessages: 144.46 KiB
BaseRealtime + decodeEncryptedMessages: 147.99 KiB
generateRandomKey: 73.80 KiB
generateRandomKey + Crypto: 73.80 KiB
getDefaultCryptoParams: 73.80 KiB
getDefaultCryptoParams + Crypto: 73.80 KiB
decodeMessage: 70.61 KiB
decodeEncryptedMessage: 74.13 KiB
decodeEncryptedMessage + Crypto: 74.13 KiB
decodeMessages: 70.71 KiB
decodeEncryptedMessages: 74.23 KiB
decodeEncryptedMessages + Crypto: 74.23 KiB
```

Resolves #1396.

### CODE REVIEW

File: src/platform/web/index.ts [hunk 1]
```diff
@@ -27,7 +26,9 @@ Platform.Config = Config;
 Platform.Transports = Transports;
 Platform.WebStorage = WebStorage;
 
-BaseClient.Crypto = Crypto;
+for (const clientClass of [DefaultRest, DefaultRealtime]) {
+  clientClass.Crypto = Crypto;
```
Comment (owenpearson) [inline src/platform/web/index.ts:30]: Would it not make more sense to just export `Crypto` separately from v2 onwards?

**Review by owenpearson — CHANGES_REQUESTED** [review]
Looks good overall, one question though:

File: src/platform/web/index.ts (hunk 1 above)
Comment (lawrence-forooghian) [reply src/platform/web/index.ts:30]: How would the non-tree-shakable version of the library then access crypto functionality?

Comment (owenpearson) [reply src/platform/web/index.ts:30]: I was thinking in `defaultrest.ts` and `defaultrealtime.ts` just import the crypto class directly?

Comment (lawrence-forooghian) [reply src/platform/web/index.ts:30]: The crypto class varies by platform, though — how would those files know which one to import?

Comment (owenpearson) [reply src/platform/web/index.ts:30]: ah yep you're right, okay let's leave it like this then 👍

**Review by owenpearson — APPROVED** [review]

**Pull request merged**